crates/red_knot_python_semantic/resources/mdtest/statically_known_branches.md

```diff
@@ -0,0 +1,1457 @@
+greater_equals_311
+
+# no error
+greater_equals_309
+
+# error: [unresolved-reference]
+less_than_309
+
+# error: [unresolved-reference]
+python2
+```
+
+### `sys.platform`
+
+```toml
+[environment]
+python-platform = "linux"
+```
+
+```py
+import sys
+
+if sys.platform == "linux":
+    linux = True
+elif sys.platform == "darwin":
+    darwin = True
+else:
+    other = True
+
+# no error
+linux
+
+# error: [unresolved-reference]
+darwin
+
+# error: [unresolved-reference]
+other
+```
+
+### `typing.TYPE_CHECKING`
+
+```py
+import typing
+
+if typing.TYPE_CHECKING:
+    type_checking = True
+else:
+    runtime = True
+
+# no error
+type_checking
+
+# error: [unresolved-reference]
+runtime
+```
+
+### Combination of `sys.platform` check and `sys.version_info` check
+
+```toml
+[environment]
+python-version = "3.10"
+python-platform = "darwin"
+```
+
+```py
+import sys
+
+if sys.platform == "darwin" and sys.version_info >= (3, 11):
+    only_platform_check_true = True
+elif sys.platform == "win32" and sys.version_info >= (3, 10):
+    only_version_check_true = True
+elif sys.platform == "linux" and sys.version_info >= (3, 11):
+    both_checks_false = True
+elif sys.platform == "darwin" and sys.version_info >= (3, 10):
+    both_checks_true = True
+else:
+    other = True
+
+# error: [unresolved-reference]
+only_platform_check_true
+
+# error: [unresolved-reference]
+only_version_check_true
+
+# error: [unresolved-reference]
+both_checks_false
+
+# no error
+both_checks_true
+
+# error: [unresolved-reference]
+other
+```
+
+## Based on type inference
+
+For the the rest of this test suite, we will mostly use `True` and `False` literals to indicate
+statically known conditions, but here, we show that the results are truly based on type inference,
+not some special handling of specific conditions in semantic index building. We use two modules to
+demonstrate this, since semantic index building is inherently single-module:
+
+```py path=module.py
+class AlwaysTrue:
+    def __bool__(self) -> Literal[True]:
+        return True
+```
+
+```py
+from module import AlwaysTrue
+
+if AlwaysTrue():
+    yes = True
+else:
+    no = True
+
+# no error
+yes
+
+# error: [unresolved-reference]
+no
+```
+
+## If statements
+
+The rest of this document contains tests for various control flow elements. This section tests `if`
+statements.
+
+### Always false
+
+#### If
+
+```py
+x = 1
+
+if False:
+    x = 2
+
+reveal_type(x)  # revealed: Literal[1]
+```
+
+#### Else
+
+```py
+x = 1
+
+if True:
+    pass
+else:
+    x = 2
+
+reveal_type(x)  # revealed: Literal[1]
+```
+
+### Always true
+
+#### If
+
+```py
+x = 1
+
+if True:
+    x = 2
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+#### Else
+
+```py
+x = 1
+
+if False:
+    pass
+else:
+    x = 2
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+### Ambiguous
+
+Just for comparison, we still infer the combined type if the condition is not statically known:
+
+```py
+def flag() -> bool: ...
```

Comment:
Referencing a review comment by @AlexWaygood here: https://github.com/astral-sh/ruff/pull/14759#discussion_r1882429730

Reply:
I really like that we have parameter type inference and can use them in tests, but somehow I like the flat structure of all the tests here. If I would use `def _(flag: bool)`, then all tests that use ambiguous truthiness would have another indentation than the tests that operator on static truthiness values.

Reply:
Yeah I don't think we have to enforce a single style here, we can use our judgment about what is nicer in each specific case. There is some inherent value in variety, in that it gives us better incidental test coverage of different areas of inference.

Reply:
> at some point we'll start emitting errors on functions like this that are annotated as returning `bool` but _really_ return `None`

I agree that's a concern; it might be better to actually include a `return True`. Of course, then that's also vulnerable to a (less certain) potential future where we do some inlining / return type inference of small functions and would infer that it always returns True.

Reply:
Done.